### PR TITLE
Sync `iterate` and `replicate` for `List` and `Vec`

### PR DESCRIPTION
**Pull request by Saransh-cpp**
Corresponding `Vec.Base.iterate` -

https://github.com/agda/agda-stdlib/blob/bfa4f8bc41da10bffaf69f5786301e587ae69c83/src/Data/Vec/Base.agda#L63-L65

There is also a similar function in Idris2 - https://github.com/idris-lang/Idris2/blob/badf1e98c8812e8b54258754cf47782f61d8fb15/libs/base/Data/List.idr#L99-L102 (not exaclty the same, but similar)

### CODE REVIEW

File: src/Data/List/Base.agda
Comment (jamesmckinna) [inline src/Data/List/Base.agda:191]: I wouldn't give the argument of type `A` the name `k`... I think that stdlib convention has that as either `x` (pros and cons either way), or else `a` (but often reserved as the level of `A`... so perhaps, sadly, best avoided). 

To me, `k` is always a `Nat`, maybe a `Fin`, or even an `Integer`...

**Comment by jamesmckinna** [comment]
Question for the stdlib developers: why is the `n` index in the `Vec`-version marked as *implicit*? 'Surely' it should be an explicit argument? (Or are the only call-sites for `iterate` ones in which the index is already inferrable from the ambient context?

File: src/Data/List/Base.agda
Comment (jamesmckinna) [reply src/Data/List/Base.agda:191]: Or, for that matter, take the `s` and `z` names from `Vec.iterate`? These are, after all, operations derived from the canonical NNO structure on`Nat`...

Comment (Saransh-cpp) [reply src/Data/List/Base.agda]: Thanks for the review! I've updated the argument names to match `Vec.Base.iterate`.

**Comment by Saransh-cpp** [comment]
Also, looking at the discussion about adding proofs in #2049. Should this PR also have some proof to ensure everything is working fine? Maybe something like (if I wrote it right) -
```agda
iterate-is-replicate-if-id : ∀ (n : ℕ) (x : A) → iterate id x n ≡ replicate n x
iterate-is-replicate-if-id  ℕ.zero x = refl
iterate-is-replicate-if-id  (suc n) x = cong (_ ∷_) (iterate-is-replicate-if-id n (id x))
```

**Comment by JacquesCarette** [comment]
Yes, adding some properties at the same time as the function is a very good idea. It should have "the same" properties as those for `Vec.iterate` as well as commuting with `toList`.

**Comment by Taneb** [comment]
> Also, looking at the discussion about adding proofs in #2049. Should this PR also have some proof to ensure everything is working fine? Maybe something like (if I wrote it right) -
> 
> ```agda
> iterate-is-replicate-if-id : ∀ (n : ℕ) (x : A) → iterate id x n ≡ replicate n x
> iterate-is-replicate-if-id  ℕ.zero x = refl
> iterate-is-replicate-if-id  (suc n) x = cong (_ ∷_) (iterate-is-replicate-if-id n (id x))
> ```

I feel like this would be better the other way around, like `replicate-is-iterate` (then you don't need to mention `id` in the name)

**Comment by jamesmckinna** [comment]
@Taneb @Saransh-cpp : why not `iterate-id-is-replicate`, as it stands?

**Comment by Saransh-cpp** [comment]
This could also be `iterate-id` (similar to `map-id`), considering the - use only LHS for naming -convention.

**Comment by Saransh-cpp** [comment]
(will rebase once #2056 is merged)

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:629]: Consider `let` binding the duplicated expression `xs = replicate n x` for the sake of legibility:
```agda
lookup-replicate : ∀ n (x : A) → let xs = replicate n x in (i : Fin (length xs)) → lookup xs i ≡ x
```
(here, and in `CHANGELOG`)

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:658]: Is this not a (fairly immediate?) consequence of `length-iterate` and `take-all`?

Even if it is worth retaining as a lemma in its own right, please consider bnding `let xs = iterate f x n in ...` in the type, as in L629 above

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:662]: I'm surprised to discover that this is *not* similarly a consequence of `length-iterate` and a lemma of the form
```agda
drop-all : (n : ℕ) (xs : List A) → n ≥ length xs → drop n xs ≡ []
drop-all n       []       _ = drop-[] n
drop-all (suc n) (x ∷ xs) p = drop-all n xs (≤-pred p)
```
because such a lemma doesn't exist! Please consider adding it as part of this PR?

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:666]: Argument `n` may be bound implicit here: it's determined uniquely by the type of `i`

File: CHANGELOG.md
Comment (jamesmckinna) [inline CHANGELOG.md:2658]: For the sake of documentation, consider the redundant recoding in the type
```agda
take-iterate : take n (iterate f x {n + m}) ≡ iterate f x {n}
```

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:629]: Better, however: you know that `length (replicate n x)` is equal to `n`, so no need to consider `i : Fin (length (replicate n x))` when `i : Fin n` will suffice?

File: src/Data/Vec/Properties.agda
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:1005]: No case analysis required; actually this lemma is not required... its proof is `refl` *definitionally* by virtue of the definition of `length` for vectors. Consider omitting.

File: src/Data/Vec/Properties.agda
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:1013]: As in `CHANGELOG` above, consider adding a redundant `{n = n}` to the RHS...

Comment (jamesmckinna) [reply src/Data/Vec/Properties.agda:1013]: Is there an 'easy' analogue of `List.Properties.take-all` which you can derive this from?

File: src/Data/Vec/Properties.agda
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:1017]: Is there an 'easy' analogue of `List.Properties.drop-all` which you can derive this from?

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:666]: And better, again, if you have shown that `length (iterate f x n)` is equal to `n`, then simplify the typing of `i` to `i : Fin n`...

File: src/Data/List/Properties.agda
Comment (JacquesCarette) [reply src/Data/List/Properties.agda:658]: The lemma is probably worth retaining, but I agree that it would be nice to have the proof use those results instead of being done "by hand", even though the resulting proof is quite simple.

File: src/Data/List/Properties.agda
Comment (Saransh-cpp) [reply src/Data/List/Properties.agda:629]: Replacing `i : Fin (length (replicate n x))` with `i: Fin n` gives me this -
```
n != (foldr (const suc) 0 (replicate n x)) of type ℕ
when checking that the expression i has type
Fin (length (replicate n x))
```

Comment (JacquesCarette) [reply src/Data/List/Properties.agda:629]: I think you'd need
```agda
lookup-replicate : ∀ n (x : A) (i : Fin n) → lookup (replicate n x) (cast (sym (length-replicate n x)) i) ≡ x
```
which does not really feel like a large improvement?

[The problem really stems from the type of `lookup` which has a computation in it that doesn't reduce, and infects everything else.]

Comment (Saransh-cpp) [reply src/Data/List/Properties.agda:629]: Thanks! I'll update this and `lookup-iterate`.

File: src/Data/List/Properties.agda
Comment (Saransh-cpp) [reply src/Data/List/Properties.agda:658]: Should it be written like this instead? -
```agda
drop-iterate : ∀ n f (x : A) → drop n (iterate f x n) ≡ []
drop-iterate n f x = drop-all n (iterate f x n) {!(length-iterate n f x)!}
```

**Comment by Saransh-cpp** [comment]
Thanks for the reviews, @jamesmckinna and @JacquesCarette!

Also, I found 2 similar proofs named very differently while going through the files, should both of them have the same name? -
```agda
take-drop-id : ∀ (m : ℕ) (x : Vec A (m + n)) → take m x ++ drop m x ≡ x
take++drop   : ∀ n (xs : List A) → take n xs ++ drop n xs ≡ xs
```

**Comment by jamesmckinna** [comment]
My feeling is that they should have the same name. What I'm not clear about us which one or even a new one, hybridising the two, such as: 
`take++drop=id`...

I'll think about it, but also throw it out to @MatthewDaggitt @taneb @gallais for their opinion(s)...

File: src/Data/Vec/Base.agda
Comment (MatthewDaggitt) [inline src/Data/Vec/Base.agda:266]: Don't need the outermost brackets on the LHS?

**Review by MatthewDaggitt — COMMENTED** [review]
> Question for the stdlib developers: why is the n index in the Vec-version marked as implicit? 'Surely' it should be an explicit argument? Or are the only call-sites for iterate ones in which the index is already inferrable from the ambient context?

I agree. The argument to `Vec.iterate` should be explicit, not implicit. @Saransh-cpp any chance you could change it as part of this PR and add a suitable CHANGELOG entry?

> Also, I found 2 similar proofs named very differently while going through the files, should both of them have the same name? -

Yup, agreed. I dislike `take-drop-id` which is very misleading. I like @jamesmckinna's suggestion of `take++drop≡id`

File: src/Data/Vec/Base.agda
Comment (Saransh-cpp) [reply src/Data/Vec/Base.agda:266]: This is coming from #2056. I will rebase once that is merged. (Some of the proofs in this PR depend on #2056.)

**Comment by Saransh-cpp** [comment]
> I agree. The argument to Vec.iterate should be explicit, not implicit

Should that be the case with `Vec.replicate` too?

**Comment by MatthewDaggitt** [comment]
> Should that be the case with Vec.replicate too?

Yes, nice spot!

**Comment by Saransh-cpp** [comment]
Re: the naming of `iterate-id` (`iterate-is-replicate`, `iterate-id-replicate`, ...) - I found -

https://github.com/agda/agda-stdlib/blob/104125c62d88ebe91274ebde521b428bad71f26e/src/Data/Fin/Properties.agda#L263-L268

which looks like a similar situation.

File: CHANGELOG.md
Comment (gallais) [inline CHANGELOG.md:2720]: Given the breaking change, I assume these `n` should now be explicit?

```suggestion

  length-iterate : length (iterate f x n) ≡ n
  iterate-id     : iterate id x n ≡ replicate x
  take-iterate   : take n (iterate f x (n + m)) ≡ iterate f x n
  drop-iterate   : drop n (iterate f x n) ≡ []
  lookup-iterate : lookup (iterate f x n) i ≡ ℕ.iterate f x (toℕ i)
  toList-iterate : toList (iterate f x n) ≡ List.iterate f x n

```

File: src/Data/Vec/Base.agda
Comment (jamesmckinna) [reply src/Data/Vec/Base.agda:264]: Nor on the RHS... ;-)

File: CHANGELOG.md
Comment (Saransh-cpp) [reply CHANGELOG.md:2720]: Yes, thanks!

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:653]:
```agda
take-iterate : ∀ n {f} {x : A} → let xs = iterate f x n in take n xs ≡ xs
take-iterate n {f} {x} = take-all n xs n≥length[xs]
  where
  xs = iterate f x n
  n≥length[xs] : n ≥ length xs
  n≥length[xs] rewrite length-iterate n {f} {x} = ≤-refl
```
NB You are going to have a hard time inferring `f` in these proofs (Agda can't unify functions easily), so, as here, the arguments will (probably: CHECK!) need to be passed explicitly, so nothing to be gained by keeping them implicit here?

**Review by jamesmckinna — CHANGES_REQUESTED** [review]
```agda
take-iterate : ∀ n {f} {x : A} → let xs = iterate f x n in take n xs ≡ xs
take-iterate n {f} {x} = take-all n xs n≥length[xs]
  where
  xs = iterate f x n
  n≥length[xs] : n ≥ length xs
  n≥length[xs] rewrite length-iterate n {f} {x} = ≤-refl
```
NB You are going to have a hard time inferring `f` in these proofs (Agda can't unify functions easily), so, as here, the arguments will (probably: CHECK!) need to be passed explicitly, so nothing to be gained by keeping them implicit here?

Similarly, it's annoying to have to repeat the `let` binding in the type and then the proof, so better to use an anonymous `module`:
```agda
module _ n f x where
  private
    xs = iterate f x n
    n≥length[xs] : n ≥ length xs
    n≥length[xs] rewrite length-iterate n {f} {x} = ≤-refl
    
  take-iterate : take n xs ≡ xs
  take-iterate = take-all n xs n≥length[xs]
```
(and then the other proofs can also be in the scope of that module...?)

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:653]: See my review comment, requesting changes, below.

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:864]: Apologies if I am the source of this `import` here! But I don't think it's necessary (there's already an import of `Base` at the top of the `Properties` module)... and may be causing the failing test?

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:875]: I think it's a good heuristic to have the parameters to a function proceed from left-to-right from 'what stays fixed' through to 'what varies throughout' (I learnt this from @JacquesCarette ... ;-)), moreover in a way which correlates with the corresponding argument order in the function invocation `iterate f x n`. So apologies are in order for not reflecting that in the module parameter list above... where `f` and `x` should come before `n`.

Indeed, within the scope of a given `f` and `x`, the module can/should consider *all* the proofs about `iterate f x` for various arguments `n`, which may vary internally (or else bind/be bound by the types of arguments like `i`), or which may be fixed as in `take-iterate` etc. 

So a bit of gentle refactoring is in order here, I think, to get the maximum value out of working in a module scope where `f` and `x` are 'fixed'...

File: src/Data/Vec/Base.agda
Comment (jamesmckinna) [inline src/Data/Vec/Base.agda:12]: Do you need the qualified import? If you do, convention/style dictates it should be imported as either `Nat` or `ℕ`.

File: src/Data/Vec/Properties.agda
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:1077]: `x` before `i` in the parameter telescope...

File: src/Data/List/Properties.agda
Comment (Saransh-cpp) [reply src/Data/List/Properties.agda]: Thanks @jamesmckinna! I'll try to refactor all the iterate proofs into this module. I've rearranged the arguments in all the proofs.

**Comment by MatthewDaggitt** [comment]
@Saransh-cpp this is no longer blocked on https://github.com/agda/agda-stdlib/pull/2056 :+1:

**Comment by Saransh-cpp** [comment]
Thanks for the ping, @MatthewDaggitt!

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:875]: Looks fine now. A possible future refactoring would be to try to squeeze the maximum out of the module abstraction across all, or most of, the lemmas. But `iterate` changes its base `x` (and counter `n`!) on each iteration, so perhaps that harder than it first seems. Ah well!

**Review by jamesmckinna — APPROVED** [review]

File: src/Data/List/Properties.agda
Comment (Saransh-cpp) [reply src/Data/List/Properties.agda:875]: Thanks, @jamesmckinna!

**Comment by MatthewDaggitt** [comment]
In the interests of getting this merged in, I've removed `take-iterate` and `drop-iterate` as I think they're probably derivative enough space cases that they don't make the bar to get included. Will merge once tests pass.

As always, thanks for your work @Saransh-cpp !

**Pull request merged**
